osquery/events/eventsubscriberplugin.cpp

```diff
@@ -635,61 +636,75 @@ void EventSubscriberPlugin::expireEventBatches(Context& context,
   }
 }
 
-EventIndex::iterator EventSubscriberPlugin::generateRows(
+EventSubscriberPlugin::GenerateRowsResult EventSubscriberPlugin::generateRows(
     Context& context,
     IDatabaseInterface& db_interface,
     std::function<void(Row)> callback,
     EventTime start_time,
     EventTime end_time,
     EventID last_eid) {
-  auto last = context.event_index.end();
-  if (end_time != 0 && start_time > end_time) {
-    return last;
-  }
+  EventSubscriberPlugin::GenerateRowsResult ret{true, 0, 0};
+  std::vector<EventID> collected_event_id_list;
+  {
+    ReadLock lock(context.event_index_mutex);
+    auto last = context.event_index.end();
+    if (end_time != 0 && start_time > end_time) {
+      return ret;
```

Comment:
I am a bit confused here, shouldn't this return `last` one, i.e. `return GenerateRowsResult{true, last->first, last->second.back()};`

Reply:
No we shouldn't. This condition checks if the start time is greater than the end time, so result will always be empty. In this case `last` iterator will still be equal to `end()` as it was assigned one line above, so it has no value.
And it doesn't really matter what we return in `last_time` and `last_eid` here, because setting `isEnd = true` indicates empty selection, and these fields will not be used.

Same applies to the next early return on line 662 when `start_time` is bigger than the latest event we have in `context.event_index`, so no events happen to be in requested interval.